### PR TITLE
initial commit for dns capture

### PR DESCRIPTION
**Pull request by PerforMance308**
Link to issue: https://github.com/kmesh-net/kmesh/issues/966

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+endif
+
+TAG ?= $(shell git rev-parse --verify HEAD)
+ifeq ($(TAG),)
+  $(error "TAG cannot be empty")
+endif
+
+.PHONY: build
+build:
+	CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -o out/$(TARGET)
```
Comment (hzxuzhonghu) [inline Makefile]: donot hardcode GOOS=linux GOARCH=amd64, we can also build for arm64

File: README.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+chmod +x ./kubectl ./kind
+sudo mv ./kubectl ./kind /usr/local/bin
+```
+
+## Step 2: Create and Activate kind
+```sh
+kind create cluster --name kmesh-dns
+```
+
+## Step3: Install Istio
```
Comment (hzxuzhonghu) [inline README.md]: seems we donot need to specify the install step 1 ~ step 3

File: README.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+kubectl apply -f manifest/deploy.yaml
+```
+
+## Step5: Forward coreDNS to Kmesh coredns plugin
+```sh
+kubectl edit cm coredns -n kube-system
+```
+
+```yaml
+apiVersion: v1
```
Comment (hzxuzhonghu) [inline README.md:18]: Instead of editing this, is there a way to create an additional cm?

File: README.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        prometheus :9153
+        forward . /etc/resolv.conf {
+           max_concurrent 1000
+        }
+        cache 30
+        loop
+        reload
+        loadbalance
+    }
+    // put the domain surfix here
```
Comment (hzxuzhonghu) [inline README.md]:
```suggestion
    // put the domain suffix here
```

File: README.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    // e.g. for consul service 
+    consul.local:53 {
+        errors
+        cache 30
+        // the KMESHPLUGIN_IP is the kmesh coredns plugin service clusterip
+        forward . $KMESHPLUGIN_IP:15053
```
Comment (hzxuzhonghu) [inline README.md:47]: does it support hostname?

File: manifest/deploy.yaml [hunk 6]
```diff
@@ -0,0 +1,47 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: kmeshdns
```
Comment (hzxuzhonghu) [inline manifest/deploy.yaml]:
```suggestion
  name: kmesh-dns
```

Comment (hzxuzhonghu) [reply manifest/deploy.yaml]: and all the below should be replaced

File: pkg/ads/ads.go [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+}
+
+func (c *AdsController) Stop() {
+	c.adsClient.Close()
+}
+
+func configInitialRequests() []*discovery.DiscoveryRequest {
+	out := make([]*discovery.DiscoveryRequest, 0, 1)
+	out = append(out, &discovery.DiscoveryRequest{
+		TypeUrl: collections.ServiceEntry.GroupVersionKind().String(),
```
Comment (hzxuzhonghu) [inline pkg/ads/ads.go:93]: Istiod doesnot respect this

**Review by hzxuzhonghu — COMMENTED** [review]
I just reviewed part of this

File: README.md (hunk 5 above)
Comment (PerforMance308) [reply README.md:47]: It only supports ip

File: README.md (hunk 2 above)
Comment (PerforMance308) [reply README.md]: yes, I will remove that if we assume that kmesh has installed

File: pkg/ads/ads.go (hunk 7 above)
Comment (PerforMance308) [reply pkg/ads/ads.go:93]: any suggestions for this, I did test it and works as expected

File: README.md (hunk 3 above)
Comment (PerforMance308) [reply README.md:18]: coreDNS seems only read config from this configmap, we can create a new one, but we have to use k8s Kustomize merge two configmaps together, but it's getting more complicated

File: pkg/ads/ads.go (hunk 7 above)
Comment (hzxuzhonghu) [reply pkg/ads/ads.go:93]: maybe i misread it. Will take a look later.

File: pkg/ads/ads.go [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+		vip:    options.GetConfig().VIP,
+	}
+	client, err := adsc.New(options.GetConfig().XDSAddress, &adsc.ADSConfig{
+		InitialDiscoveryRequests: configInitialRequests(),
+		ResponseHandler:          c,
+		Config: adsc.Config{
+			Workload:  options.GetConfig().ServiceNode,
+			Namespace: options.GetConfig().ServiceNameSpace,
+			Meta: model.NodeMetadata{
+				Generator: "api",
```
Comment (hzxuzhonghu) [inline pkg/ads/ads.go:59]: IC, this is setting the generator

File: manifest/deploy.yaml [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    matchLabels:
+      app: kmesh-dns
+  template:
+    metadata:
+      labels:
+        app: kmesh-dns
+    spec:
+      containers:
+      - name: kmesh-dns
+        image: ghcr.io/kmesh-net/kmesh-dns:f436b36ea43c0b013358959d83dc866de7de17d7
```
Comment (hzxuzhonghu) [inline manifest/deploy.yaml]:
```suggestion
        image: ghcr.io/kmesh-net/kmesh-dns:latest
```

File: pkg/ads/ads.go [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+	c.adsClient.Close()
+}
+
+func configInitialRequests() []*discovery.DiscoveryRequest {
+	out := make([]*discovery.DiscoveryRequest, 0, 1)
+	out = append(out, &discovery.DiscoveryRequest{
+		TypeUrl: collections.ServiceEntry.GroupVersionKind().String(),
+	})
+
+	return out
```
Comment (hzxuzhonghu) [inline pkg/ads/ads.go]:
```suggestion
	return []*discovery.DiscoveryRequest{&discovery.DiscoveryRequest{
		TypeUrl: collections.ServiceEntry.GroupVersionKind().String(),
	}}
	
```

File: pkg/ads/ads.go [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+func (c *AdsController) HandleResponse(con *adsc.ADSC, response *discovery.DiscoveryResponse) {
+	dnsEntries := make(map[string][]net.IP)
+	for _, r := range response.Resources {
+		m := &mcp.Resource{}
+		if err := anypb.UnmarshalTo(r, m, proto.UnmarshalOptions{}); err != nil {
+			fmt.Println("Error unmarshalling:", err)
+			continue
+		}
+		e := &networking.ServiceEntry{}
+		if err := anypb.UnmarshalTo(m.Body, e, proto.UnmarshalOptions{}); err != nil {
```
Comment (hzxuzhonghu) [inline pkg/ads/ads.go]: We can use ADSC.XDSUpdates to get the resp without being aware of mcp.Resource

File: pkg/ads/ads.go [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+		if e.Resolution == networking.ServiceEntry_NONE {
+			// NO DNS based service discovery for service entries
+			// that specify NONE as the resolution. NONE implies
+			// that Istio should use the IP provided by the caller
+			continue
+		}
+		addresses := e.Addresses
+		if len(addresses) == 0 && c.vip != "" {
+			// If the ServiceEntry has no Addresses, map to a user-supplied default value, if provided
+			addresses = []string{c.vip}
```
Comment (hzxuzhonghu) [inline pkg/ads/ads.go:119]: what does vip mean

File: pkg/ads/ads.go [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+		if len(vips) == 0 {
+			continue
+		}
+		for _, host := range e.Hosts {
+			key := fmt.Sprintf("%s.", host)
+			if strings.Contains(host, "*") {
+				// Validation will ensure that the host is of the form *.foo.com
+				parts := strings.SplitN(host, ".", 2)
+				// Prefix wildcards with a . so that we can distinguish these entries in the map
+				key = fmt.Sprintf(".%s.", parts[1])
```
Comment (hzxuzhonghu) [inline pkg/ads/ads.go:131]: I am not an expert on this, is there a rfc to refer.

Comment (hzxuzhonghu) [reply pkg/ads/ads.go:131]: I think parts[1] can cause panic

File: pkg/ads/ads.go [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+			}
+			dnsEntries[key] = vips
+		}
+	}
+	c.server.UpdateDNSEntries(dnsEntries)
+
+	// send ACK
+	ackRequest := &discovery.DiscoveryRequest{
```
Comment (hzxuzhonghu) [inline pkg/ads/ads.go]: we donot need to send ack manually, adsc send it for us already

File: pkg/dns/dns.go [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+		addr = "localhost:15053"
+	}
+	host, port, err := net.SplitHostPort(addr)
+	if err != nil {
+		return nil, fmt.Errorf("dns address must be a valid host:port: %v", err)
+	}
+	addresses := []string{addr}
+	if host == "localhost" {
+		addresses = []string{}
+		addresses = append(addresses, net.JoinHostPort("0.0.0.0", port))
```
Comment (hzxuzhonghu) [inline pkg/dns/dns.go]: ? should you remove local host? If address == "" or address == ":15053", or `0.0.0:15053`, make them listen on 0.0.0:15053

File: pkg/dns/dns.go [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+	}
+
+	for _, ipAddr := range addresses {
```
Comment (hzxuzhonghu) [inline pkg/dns/dns.go]: I cannot see there can be multiple addresses from the above implement

File: pkg/server.go [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+	}
+
+	m.ads = adsController
+
+	return m, nil
+}
+
+func (m *Manager) Start(stop <-chan struct{}) error {
+	m.server.Start()
+	if err := m.ads.Start(); err != nil {
```
Comment (hzxuzhonghu) [inline pkg/server.go]: should pass the stop channel down

File: test/run_test.sh
Comment (hzxuzhonghu) [inline test/run_test.sh:1]: good to see e2e test, can we enable it in ci. Not a blocker, can be done in the following pr

File: pkg/ads/ads.go (hunk 12 above)
Comment (PerforMance308) [reply pkg/ads/ads.go:119]: VIP refers to `default value for A record, if ServiceEntry has no Addresses`. If no address is specified in the ServiceEntry, the CoreDNS plugin will use this default VIP to resolve the domain name. This ensures that even if there is no IP address, the domain name can still be resolved to a virtual address, avoiding resolution failures.

File: pkg/ads/ads.go (hunk 13 above)
Comment (PerforMance308) [reply pkg/ads/ads.go:131]: This piece of code is actually from Istio coredns plugin: https://github.com/istio-ecosystem/istio-coredns-plugin/blob/b84dba7e1aff4e29d4074188b58d2189edeb2573/plugin.go#L101

File: pkg/ads/ads.go (hunk 11 above)
Comment (PerforMance308) [reply pkg/ads/ads.go]: What's the impact of current implementation？Do we must be not aware of mcp.Resource, or it is ok just keep current implementation

File: test/run_test.sh
Comment (PerforMance308) [reply test/run_test.sh:1]: will add it in the following pr

File: pkg/ads/ads.go [hunk 18]
```diff
@@ -14,154 +14,146 @@
...
-		vip:    options.GetConfig().VIP,
-	}
-	client, err := adsc.New(options.GetConfig().XDSAddress, &adsc.ADSConfig{
-		InitialDiscoveryRequests: configInitialRequests(),
-		ResponseHandler:          c,
-		Config: adsc.Config{
-			Workload:  options.GetConfig().ServiceNode,
-			Namespace: options.GetConfig().ServiceNameSpace,
-			Meta: model.NodeMetadata{
-				Generator: "api",
```
Comment (hzxuzhonghu) [inline pkg/ads/ads.go]: why remove it?

File: pkg/ads/ads.go [hunk 19]
```diff
@@ -14,154 +14,146 @@
...
+				 // Because we use the custom grpc options for adsc, here we should
+				 // explicitly set transport credentials.
+				 // TODO: maybe we should use the tls settings within ConfigSource
+				 // to secure the connection between istiod and remote xds server.
+				 grpc.WithTransportCredentials(insecure.NewCredentials()),
+			 },
+		 },
+	 })
+	 if err != nil {
+		 log.Fatalf("failed to dial XDS %s %v", "controller.gateway-system.svc.cluster.local:15010", err)
```
Comment (hzxuzhonghu) [inline pkg/ads/ads.go]: prevent hardcoding, and i think controller.gateway-system.svc.cluster.local:15010 is not right

File: pkg/ads/ads.go [hunk 20]
```diff
@@ -14,154 +14,146 @@
...
+ 
+	 go func() {
+		 <-stop
+		 c.Stop()
+	 }()
+	 return nil
+ }
+ 
+ func (c *AdsController) Stop() {
+	 c.adsClient.Close()
```
Comment (hzxuzhonghu) [inline pkg/ads/ads.go]: can we move this into `Start`, seems this function is not needed

File: pkg/ads/ads.go [hunk 21]
```diff
@@ -14,154 +14,146 @@
...
-				vips = append(vips, ip)
-			}
-		}
-	}
-
-	return vips
-}
+ import (
+	 "fmt"
+	 "log"
```
Comment (hzxuzhonghu) [inline pkg/ads/ads.go]: nit: we donot use this package

File: pkg/ads/ads.go [hunk 22]
```diff
@@ -14,154 +14,146 @@
...
+		 TypeUrl: collections.ServiceEntry.GroupVersionKind().String(),
+	 }}
+ }
+ 
+ func (c *AdsController) HandleResponse(con *adsc.ADSC, response *discovery.DiscoveryResponse) {
+	 dnsEntries := make(map[string][]net.IP)
+	 for _, r := range response.Resources {
+		 m := &mcp.Resource{}
+		 if err := anypb.UnmarshalTo(r, m, proto.UnmarshalOptions{}); err != nil {
+			 fmt.Println("Error unmarshalling Resource:", err)
```
Comment (hzxuzhonghu) [inline pkg/ads/ads.go]: nit:  we should use kmesh logger in the long run, i donot mind using this at first version

**Review by hzxuzhonghu — COMMENTED** [review]
I will approve once  the comments addressed, and i would suggest test this manually again

File: pkg/ads/ads.go (hunk 18 above)
Comment (PerforMance308) [reply pkg/ads/ads.go]: I thought your comment before said this is not necessary, will add it back

**Review by hzxuzhonghu — APPROVED** [review]
/lgtm

**Pull request merged**